### PR TITLE
Emit terminate event when runner has completed.

### PR DESCRIPTION
**Pull request by johndagostino**
This allows the parent process to listen to the runner to determine if it has successfully completed processing all messages.

This lives along side the terminate method that already has a terminateWait behaviour.

Introduce a runner state that shows the current state of the runner.

This allows us to wait until the processing is completed before disconnect completes.

### CODE REVIEW

File: src/runner/kafka.ts [hunk 1]
```diff
@@ -294,7 +296,7 @@ class KafkaRunner extends BaseRunner
     if (!this.consumer.isConnected()) {
       throw new Error('Lost connection to kafka');
     }
-    if (!this.steveo.paused) {
+    if (this.state === 'paused') {
```
Comment (arshu16) [inline src/runner/kafka.ts:299]: `!==` paused and terminating

File: src/index.ts [hunk 2]
```diff
@@ -262,9 +262,9 @@ export class Steveo implements ISteveo {
     return this._runner;
   }
 
-  disconnect() {
-    this._producer?.disconnect();
-    this._runner?.disconnect();
+  async disconnect() {
```
Comment (sugendran) [inline src/index.ts:265]: Should this be a `Promise.all` so that it's not sequential?

File: package.json [hunk 3]
```diff
@@ -1,15 +1,15 @@
 {
   "name": "steveo",
-  "version": "5.0.0-beta8",
+  "version": "5.0.0-beta9",
```
Comment (rrsmith) [inline package.json:3]: Major when released I think given we're changing the behaviour of terminate

**Review by rrsmith — APPROVED** [review]
Ishs comment r.e. the resume behaviour looks to be the only functional issue that needs sorting

File: src/registry.ts [hunk 4]
```diff
@@ -20,7 +20,7 @@ class Registry implements IRegistry {
 
   emit(name: string, ...args: any) {
     this.heartbeat = Math.max(new Date().getTime(), this.heartbeat);
-    this.emit(name, ...args);
+    this.events.emit(name, ...args);
```
Comment (sugendran) [inline src/registry.ts:23]: Given you're fixing this function was calling itself, maybe it's a good time to introduce a test for `emit`

File: src/runner/redis.ts [hunk 5]
```diff
@@ -134,7 +134,12 @@ class RedisRunner extends BaseRunner implements IRunner {
 
   async process(topics?: string[]) {
     const loop = () => {
```
Comment (sugendran) [inline src/runner/redis.ts:136]: Why is loop redefined on every call to `process` ?

**Pull request merged**